sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,359 @@
+<h2>Array Parameters of SYCL Kernels</h2>
+
+<h3>Introduction</h3>
+
+This document describes the changes to support passing arrays to SYCL kernels and special treatment of Accessor arrays. The following cases are handled:
+
+1. arrays of standard-layout type as top-level arguments
+2. arrays of Accessors as top-level arguments
+3. arrays of accessors within structs that are top-level arguments
+
+The motivation for this correction to kernel parameters processing is to bring consistency to the treatment of arrays. On the CPU, a lambda function is allowed to access an element of an array defined outside the lambda. The implementation captures the entire array by value. A user would naturally expect this to work in SYCL as well. However, the current implementation flags references to arrays defined outside a SYCL kernel as errors.
+
+The first few sections describe the current design. The last three sections describe the design to support 1. to 3. above. The implementation of this design is confined to three functions in the file SemaSYCL.cpp.
+
+<h3>A SYCL Kernel</h3>
+
+The SYCL constructs single_task, parallel_for, and parallel_for_work_group each take a function object or a lambda function as one of their arguments. The code within the function object or lambda function is executed on the device. Code generation for SYCL is based on the internal representation of the function/lambda object.
+
+<h3>SYCL Kernel Code Generation</h3>
+
+Consider a source code example that captures an int, a struct and an accessor by value:
+
+```
+constexpr size_t c_num_items = 10;
+range<1> num_items{c_num_items}; // range<1>(num_items)
+
+int main()
+{
+  int output[c_num_items];
+  queue myQueue;
+
+  int i = 55;
+  struct S {
+    int m;
+  } s = { 66 };
+  auto outBuf = buffer<int, 1>(&output[0], num_items);
+
+  myQueue.submit([&](handler &cgh) {
+    auto outAcc = outBuf.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class Worker>(num_items, [=](cl::sycl::id<1> index) {
+      outAcc[index] = i + s.m;
+    });
+  });
+
+  return 0;
+}
+```
+
+The input to the code generation routines is a function object that represents the kernel. In pseudo-code:
+
+```
+struct Capture {
+    Accessor outAcc;
+    int i;
+    struct S s;
+    () {
+        outAcc[index] = i + s.m;
+    }
+}
+```
+
+On the CPU a call to such a lambda function would look like this:
+`()(struct Capture* this);`
+       
+When offloading the kernel to a device, the lambda function cannot be directly called with a capture object address. Instead, the code generated for the device is in the form of a �kernel caller� and a �kernel callee�. The callee looks very similar to a lambda function on the CPU. The caller requires special code generation. It is designed to receive the lambda capture object in pieces, assemble the pieces into the original lambda capture object and then call the callee:

```

Comment:
Right, the "kernel caller" is generated OpenCL kernel function. The "kernel callee" is lambda object taken from user code as is. We actually don't generate lambda call in SemaSYCL, just reuse special code from SYCL headers. SemaSYCL generates only initialization of local copy of lambda.